tests/runner/check_serial_runner_memory.py

```diff
@@ -0,0 +1,139 @@
+        self.items = [x for x in range(MEMORY_EATER_LIST_SIZE)]
+
+    def stop_node(self, node):
+        pass
+
+    def clean_node(self, node):
+        pass
+
+
+class MemoryLeakTest(Test):
+    """A group of identical "memory-hungry" ducktape tests.
+    Each test holds a reference to a service which itself holds a reference to a large (memory intensive) object.
+    """
+    def __init__(self, test_context):
+        super(MemoryLeakTest, self).__init__(test_context)
+        self.memory_eater = MemoryEater(test_context)
+
+    @matrix(x=[i for i in range(N_TEST_CASES)])
+    def test_leak(self, x):
+        self.memory_eater.start()
+
+
+class InstrumentedSerialTestRunner(SerialTestRunner):
+    """Identical to SerialTestRunner, except dump memory used by the current process
+    before running each test.
+    """
+    def __init__(self, *args, **kwargs):
+        self.memory_data_file = kwargs.get("memory_data_file")
+        del kwargs["memory_data_file"]
+        super(InstrumentedSerialTestRunner, self).__init__(*args, **kwargs)
+
+    def run_single_test(self):
+
+        # write current memory usage to file before running the test
+        pid = os.getpid()
+        current_memory = _get_memory(pid)
+        with open(self.memory_data_file, "a+") as f:
+            f.write(str(current_memory) + "\n")
+
+        data = super(InstrumentedSerialTestRunner, self).run_single_test()
+        return data
+
+
+class CheckMemoryUsage(object):
+    def setup_method(self, _):
+        mock_cluster = LocalhostCluster()
+        self.session_context = tests.ducktape_mock.session_context(mock_cluster)
+        self.tmp_dir = tempfile.mkdtemp()
+        self.tmp_file = os.path.join(self.tmp_dir, "tmp_file")
+
+    def check_for_inter_test_memory_leak(self):
+        """Until v0.3.10, ducktape had a serious source of potential memory leaks.
+
+        Because test_context objects held a reference to all services for the duration of a test run, the memory
+        used by any individual service would not be garbage-collected until well after *all* tests had run.
+
+        This memory leak was discovered in Kafka system tests, where many long-running system tests were enough
+        to cumulatively use up the memory on the test machine, causing a cascade of test failures due to
+        inability to allocate any more memory.
+
+        This test provides a regression check against this type of memory leak; it fails without the fix, and passes
+        with it.
+        """
+        # Get a list of test_context objects for the test runner
+        ctx_list = []
+        test_methods = [MemoryLeakTest.test_leak]
+        for f in test_methods:
+            ctx_list.extend(MarkedFunctionExpander(session_context=self.session_context, cls=MemoryLeakTest, function=f).expand())
+        assert len(ctx_list) == N_TEST_CASES  # Sanity check
+
+        runner = InstrumentedSerialTestRunner(self.session_context, ctx_list, memory_data_file=self.tmp_file)
```

Comment:
Yeah it was purely for sharing data between processes - but there are simpler ways to do this.